### PR TITLE
Fix skipped logic and wrong casts for fp16/bf16 for fused sgd cpu

### LINKED ISSUE #191763: Fused SGD on CPU silently skips vectorized fp16 and bf16 parameter updates in certain circumstances
### 🐛 Describe the bug

The optimizer `torch.optim.SGD` fails to update the vectorized parts of fp16 and bf16 params when numel() >= SIMD width (SIMD width = `lpVec::size()` / `at::vec::Vectorized<scalar_t>::size()` in the CPU fused-SGD kernel, and that == 8 on my system) and fused=True.  Works fine for fp32 and fp64, only the low precision types have the problem.

- when numel() < SIMD width, the angels sing and world peace begins because it works
- when numel() % SIMD width == 0, the param update is a no-op
- when numel() > SIMD width, every complete SIMD block isn't touched, and only the scalar remainder updates

PyTorch internal testing does not find this bug because it's model has small params, sizes 6, 3, 3, and 1, and that isn't going to be >= SIMD width on any hardware I know of, and the existing large tensor fused optimizer test uses zero gradients so the result is expected to not have changed params, and hence doesn't expose the issue.

## Reproducing the issue

I reproduced this on my m3 max with both PyTorch 2.12.1 and 2.13.0.  

Code to reproduce:
```python
import platform
import torch

def run_step(dtype: torch.dtype, size: int, *, fused: bool) -> torch.Tensor:
    parameter = torch.nn.Parameter(torch.ones(size, dtype=dtype, device="cpu"))
    parameter.grad = torch.ones_like(parameter)

    torch.optim.SGD([parameter], lr=0.125, fused=fused).step()
    return parameter.detach()


def result_text(tensor: torch.Tensor) -> str:
    total = tensor.numel()
    updated = int(torch.count_nonzero(tensor == 0.875).item())
    return f"{updated} out of {total} elements updated to 0.875"


print(f"PyTorch: {torch.__version__}")
print(f"Platform: {platform.platform()}")

for dtype in (torch.float16, torch.bfloat16):
    for size in (7, 8, 9, 64, 65):
        unfused = run_step(dtype, size, fused=False)
        fused = run_step(dtype, size, fused=True)

        print(f"\n{dtype}, {size} elements")
        print(f"  unfused: {result_text(unfused)}")
        print(f"  fused:   {result_text(fused)}")
```

Output for 2.13.0
```text
PyTorch: 2.13.0
Platform: macOS-26.3-arm64-arm-64bit-Mach-O

torch.float16, 7 elements
  unfused: 7 out of 7 elements updated to 0.875
  fused:   7 out of 7 elements updated to 0.875

torch.float16, 8 elements
  unfused: 8 out of 8 elements updated to 0.875
  fused:   0 out of 8 elements updated to 0.875

torch.float16, 9 elements
  unfused: 9 out of 9 elements updated to 0.875
  fused:   1 out of 9 elements updated to 0.875

torch.float16, 64 elements
  unfused: 64 out of 64 elements updated to 0.875
  fused:   0 out of 64 elements updated to 0.875

torch.float16, 65 elements
  unfused: 65 out of 65 elements updated to 0.875
  fused:   1 out of 65 elements updated to 0.875

torch.bfloat16, 7 elements
  unfused: 7 out of 7 elements updated to 0.875
  fused:   7 out of 7 elements updated to 0.875

torch.bfloat16, 8 elements
  unfused: 8 out of 8 elements updated to 0.875
  fused:   0 out of 8 elements updated to 0.875

torch.bfloat16, 9 elements
  unfused: 9 out of 9 elements updated to 0.875
  fused:   1 out of 9 elements updated to 0.875

torch.bfloat16, 64 elements
  unfused: 64 out of 64 elements updated to 0.875
  fused:   0 out of 64 elements updated to 0.875

torch.bfloat16, 65 elements
  unfused: 65 out of 65 elements updated to 0.875
  fused:   1 out of 65 elements updated to 0.875
```

Output for 2.12.1
```text
PyTorch: 2.12.1
Platform: macOS-26.3-arm64-arm-64bit-Mach-O

torch.float16, 7 elements
  unfused: 7 out of 7 elements updated to 0.875
  fused:   7 out of 7 elements updated to 0.875

torch.float16, 8 elements
  unfused: 8 out of 8 elements updated to 0.875
  fused:   0 out of 8 elements updated to 0.875

torch.float16, 9 elements
  unfused: 9 out of 9 elements updated to 0.875
  fused:   1 out of 9 elements updated to 0.875

torch.float16, 64 elements
  unfused: 64 out of 64 elements updated to 0.875
  fused:   0 out of 64 elements updated to 0.875

torch.float16, 65 elements
  unfused: 65 out of 65 elements updated to 0.875
  fused:   1 out of 65 elements updated to 0.875

torch.bfloat16, 7 elements
  unfused: 7 out of 7 elements updated to 0.875
  fused:   7 out of 7 elements updated to 0.875

torch.bfloat16, 8 elements
  unfused: 8 out of 8 elements updated to 0.875
  fused:   0 out of 8 elements updated to 0.875

torch.bfloat16, 9 elements
  unfused: 9 out of 9 elements updated to 0.875
  fused:   1 out of 9 elements updated to 0.875

torch.bfloat16, 64 elements
  unfused: 64 out of 64 elements updated to 0.875
  fused:   0 out of 64 elements updated to 0.875

torch.bfloat16, 65 elements
  unfused: 65 out of 65 elements updated to 0.875
  fused:   1 out of 65 elements updated to 0.875
```

### Cause

In the fused SGD kernel for fp16 and bf16 the vectorized loop does the calculations it needs, but then the loop ends without applying the LR update.  Like any sane developer now days I use AI so I can be lazy, so here's what AI put together to point out where the problem is:

## Source of the problem

In the FP16/BF16 specialization of the CPU fused-SGD kernel, the
vectorized loop calculates the gradients and momentum but reaches the
end of the loop without applying the learning-rate update to
`param_vec1` and `param_vec2` or storing them back to `param_ptr`:

https://github.com/pytorch/pytorch/blob/4a4f3e8cfe754f198669b258a5e1c3f8cec0f247/aten/src/ATen/native/cpu/FusedSGDKernel.cpp#L35-L74

The scalar remainder correctly updates `param_ptr`:

https://github.com/pytorch/pytorch/blob/4a4f3e8cfe754f198669b258a5e1c3f8cec0f247/aten/src/ATen/native/cpu/FusedSGDKernel.cpp#L75-L102

The FP32/FP64 vectorized implementation also performs and stores its
parameter update:

https://github.com/pytorch/pytorch/blob/4a4f3e8cfe754f198669b258a5e1c3f8cec0f247/aten/src/ATen/native/cpu/FusedSGDKernel.cpp#L126-L156

The same missing update is still present in the current 2.14
development source. 

### Versions

PyTorch version: 2.12.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 26.3 (arm64)
GCC version: Could not collect
Clang version: 17.0.0 (clang-1700.6.3.2)
CMake version: version 4.3.4
Libc version: N/A

Python version: 3.14.2 (main, Dec  5 2025, 16:49:16) [Clang 17.0.0 (clang-1700.4.4.1)] (64-bit runtime)
Python platform: macOS-26.3-arm64-arm-64bit-Mach-O
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Apple M3 Max

Versions of relevant libraries:
[pip3] numpy==2.5.0
[pip3] torch==2.12.1
[pip3] torchcts==0.4.1
[conda] numpy                     2.5.1                    pypi_0    pypi
[conda] numpydoc                  1.7.0           py312hca03da5_0  
[conda] tbb                       2021.8.0             h48ca7d4_0  
[conda] torch                     2.6.0                    pypi_0    pypi

cc @ezyang @gchanan @kadeng @msaroufim @vincentqb @jbschlosser @albanD @janeyx99 @crcrpar @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01

### PR DESCRIPTION
**Pull request by janeyx99**
Fixes #191763 

Also fixes:
- Use opmath_t for compute and not scalar_t
- momentum path was writing trash

Adds a new test case. I was trying to fold it into existing tests, but
- fused vs forloop diverged too much due to intermediates being opmath and not scalar_t
- fused cpu vs cuda diverges cuz CPU is different for Linear than GPU
- this is the most straightforward test

Perf is not affected, if anything it became faster? 
image

Script to run before and after the change:
```
import json, sys, time
import torch

LABEL = sys.argv[1]
CASES = [
    ("Adagrad", torch.optim.Adagrad, {"lr": 0.1, "eps": 1e-10}),
    ("Adagrad_wd", torch.optim.Adagrad, {"lr": 0.1, "eps": 1e-10, "weight_decay": 0.1}),
    ("SGD_mom", torch.optim.SGD, {"lr": 0.01, "momentum": 0.9}),
    ("SGD_wd_mom", torch.optim.SGD, {"lr": 0.01, "momentum": 0.9, "weight_decay": 0.1}),
]
DTYPES = [("bf16", torch.bfloat16), ("fp16", torch.float16)]
N = 4096          # accuracy: many vector blocks plus a remainder
NPERF = 1 << 22   # perf: 4.2M elements
STEPS = 5

def accuracy(optim_cls, kwargs, dtype):
    """Intra-kernel vec-vs-scalar gap, and error against an fp64 reference."""
    # every slot identical -> the size-1 param is the exact expected value
    big = torch.nn.Parameter(torch.full((N + 1,), 0.25, dtype=dtype))
    one = torch.nn.Parameter(torch.full((1,), 0.25, dtype=dtype))
    ref64 = torch.nn.Parameter(torch.full((1,), 0.25, dtype=torch.float64))
    ob = optim_cls([big], fused=True, **kwargs)
    oo = optim_cls([one], fused=True, **kwargs)
    o64 = optim_cls([ref64], **kwargs)
    for s in range(STEPS):
        gv = 0.125 * (s + 1)
        for p, o in ((big, ob), (one, oo), (ref64, o64)):
            p.grad = torch.full_like(p, gv)
            o.step()
    vec_vs_scalar = (big.detach().double() - one.detach().double()).abs().max().item()
    err_vs_fp64 = (big.detach().double() - ref64.detach()).abs().max().item()
    return vec_vs_scalar, err_vs_fp64

def perf(optim_cls, kwargs, dtype):
    p = torch.nn.Parameter(torch.full((NPERF,), 0.25, dtype=dtype))
    o = optim_cls([p], fused=True, **kwargs)
    g = torch.full_like(p, 0.01)
    for _ in range(3):
        p.grad = g.clone()
        o.step()
    best = float("inf")
    for _ in range(7):
        p.grad = g.clone()
        t0 = time.perf_counter()
        o.step()
        best = min(best, time.perf_counter() - t0)
    return NPERF / best / 1e6

out = {}
for cname, cls, kw in CASES:
    for dname, dt in DTYPES:
        v, e = accuracy(cls, kw, dt)
        out[f"{cname}/{dname}"] = {
            "vec_vs_scalar": v,
            "err_vs_fp64": e,
            "Melem_s": perf(cls, kw, dt),
        }
print(json.dumps({"label": LABEL, "results": out}, indent=1))
```

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01

### CODE REVIEW

**Review by Skylion007 — APPROVED** [review]

File: test/test_optim.py [hunk 1]
```diff
@@ -1156,6 +1156,51 @@ def test_fused_matches_forloop(self, device, dtype, optim_info):
             )
         self._test_derived_optimizers(device, dtype, optim_info, "fused")
 
+    # Test vectorization path splits for fused kernels. Ensure the scalar path
+    # and vectorized path compute the same thing.
+    # See https://github.com/pytorch/pytorch/issues/165632.
```
Comment (albanD) [inline test/test_optim.py]: Interesting read, thanks for pointing at it!

**Review by albanD — APPROVED** [review]
SGTM if there is no major perf regression from the upcasts.

File: test/test_optim.py (hunk 1 above)
Comment (janeyx99) [inline test/test_optim.py]:
```suggestion
    # See https://github.com/pytorch/pytorch/issues/191763.
```
